chore: track size of a "Hello world" app built with WebPack

pkozlowski-opensource opened the pull request.

pkozlowski-opensource committed:
chore: track size of a "Hello world" app built with WebPack

pkozlowski-opensource commented:
This is still WIP but opening now to be able to coordinate with @IgorMinar and @yjbanov 

@IgorMinar and @yjbanov - I would like to discuss how to organize folders so both JS and Dart versions are "close" to each other but without stepping on each other toes

pkozlowski-opensource commented:
Remaining tasks:
* [ ] co-exist / unify with Dart
* [ ] run e2e tests before reporting numbers (only report numbers when tests are passing)
* [ ] make sure that we can have multiple tests (hello world, TodoMVC etc.)

yjbanov commented:
We are already measuring the size of "hello world" app under `modules/playground`. Do we want to switch to a different app?

As for directories, for Dart it would be straight-forward to reuse the `modules_dart` directory, which is already friends with gulp. What do you think?

pkozlowski-opensource commented:
> We are already measuring the size of "hello world" app under modules/playground. Do we want to switch to a different app?

Yes, this is what I got from the discussion with @IgorMinar in https://docs.google.com/document/d/14_PBtkOL0vyyRFmXxkRnz6GFWXjYisxZohDxdl1jV4c/edit#
Indeed setup for a WebPack-based app will be different as compared to a generic playground one.

> As for directories, for Dart it would be straight-forward to reuse the modules_dart directory, which is already friends with gulp. What do you think?

Yup, I saw your work on the Dart tracking, but my understanding was that @IgorMinar wants to unify somehow both TS and Dart (as much as it makes sense, of course)

yjbanov commented:
Sounds good. Looks like you have some vision behind this. I'd like to play an implementor role in this effort, so tell me what I should do :) Otherwise, I can just plug a todo app into what we currently use to measure Dart code size.

IgorMinar commented on tools/broccoli/trees/dart_tree.ts:
```diff
@@ -16,6 +16,7 @@ import replace from '../broccoli-replace';
 var global_excludes = [
   'angular2/http*',
   'angular2/upgrade*',
+  'payload_tests/**/*',
```
Alphabetize the list please

IgorMinar commented on gulpfile.js:
```diff
@@ -658,6 +660,64 @@ gulp.task('!checkAndReport.payload.dart', function() {
+
+gulp.task('!build.payload.js', function() {
+  var q = require('q');
+  var webpack = q.denodeify(require('webpack'));
```
Don't we have pure denodeify package (independent of q in our project already?) in case we do please use that instead of Q.

I'm trying to remove Q from our project completely.

pkozlowski-opensource replied:
Nope, we don't have it as a dep right now. But I wanted to change how we denodeify webpack anyway as the current version doesn't handle all the error cases correctly. Will do it in a separate PR.

IgorMinar commented on gulpfile.js:
```diff
@@ -658,6 +660,64 @@ gulp.task('!checkAndReport.payload.dart', function() {
+
+gulp.task('!checkAndReport.payload.js', function() {
+  var reportSize = require('./tools/analytics/reportsize');
+  var PAYLOAD_SIZE_LIMITS = {'uncompressed': 550 * 1024, 'gzip level=6': 120 * 1024};
```
Let's use level=9 instead. I checked with our serving gurus and they said that there is no reason not to use level=9 by default for all static assets.

IgorMinar commented on gulpfile.js:
```diff
@@ -658,6 +660,64 @@ gulp.task('!checkAndReport.payload.dart', function() {
+            gulp.src([
+                  'node_modules/zone.js/dist/zone-microtask.js',
+                  'node_modules/zone.js/dist/long-stack-trace-zone.js',
+                  'node_modules/reflect-metadata/Reflect.js',
```
Are you intentionally not using the angular2-polyfills.js?

pkozlowski-opensource replied:
Yeh, just to avoid building all the bundles.

IgorMinar commented on modules/payload_tests/src/hello_world_webpack/index.ts:
```diff
@@ -0,0 +1,15 @@
+import {Component} from 'angular2/core';
```
Maybe rename the directory from:

modules/payload_tests/src/hello_world_webpack/index.ts

To:
modules/payload_tests/hello_world/ts/webpack/index.ts

IgorMinar commented:
The rest looks good

pkozlowski-opensource commented:
@IgorMinar I've addressed most of your comments and going tackle the remaining one (removing calls to `q.denodify`) in a separate PR. Do you want to have another look at this PR or is it good to go?

mary-poppins commented:
Merging PR #6434 on behalf of @rkirov to branch [presubmit-rkirov-pr-6434](https://github.com/angular/angular/tree/presubmit-rkirov-pr-6434).

yjbanov commented on gulpfile.js:
```diff
@@ -658,6 +672,64 @@ gulp.task('!checkAndReport.payload.dart', function() {
+  return q.all(PAYLOAD_TESTS_CONFIG.ts.webpack.cases.map(function(caseName) {
+    var CASE_PATH = PAYLOAD_TESTS_CONFIG.ts.webpack.dist(caseName);
+
+    return webpack({
```
Instead of putting it in our main gulpfile.js, can the webpack code be moved closer to the application itself? This way one doesn't need to look in multiple places to figure out how to package an app for production.
